src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -1060,12 +1203,61 @@ struct BlockZipperJoinImpl {
   }
 
   // The actual join routine that combines all the previous functions.
+  // not been consumed yet and match them against undef blocks.
+  template <bool reversed>
+  bool consumeRemainingBlocks(auto& side, const auto& undefBlocks) {
+    bool hasUndef = false;
+    while (side.it_ != side.end_) {
+      const auto& lBlock = *side.it_;
+      for (const auto& rBlock : undefBlocks) {
+        if constexpr (reversed) {
+          compatibleRowAction_.setInput(rBlock.fullBlock(), lBlock);
+        } else {
+          compatibleRowAction_.setInput(lBlock, rBlock.fullBlock());
+        }
+        for (size_t i : ad_utility::integerRange(lBlock.size())) {
+          for (size_t j : rBlock.getIndexRange()) {
+            hasUndef = true;
+            if constexpr (reversed) {
+              compatibleRowAction_.addRow(j, i);
+            } else {
+              compatibleRowAction_.addRow(i, j);
+            }
+          }
+        }
+      }
+      ++side.it_;
+    }
+    return hasUndef;
+  }
+
+  // If one of the sides is exhausted and has no values to match the pairs
+  // left, we need to pair the remaining values with the undef values we have
+  // left.
+  void addRemainingUndefPairs() {
+    if constexpr (potentiallyHasUndef) {
+      addCartesianProduct(leftSide_.currentBlocks_, rightSide_.undefBlocks_);
+      consumeRemainingBlocks<false>(leftSide_, rightSide_.undefBlocks_);
+
+      addCartesianProduct(leftSide_.undefBlocks_, rightSide_.currentBlocks_);
+      consumeRemainingBlocks<true>(rightSide_, leftSide_.undefBlocks_);
+
+      compatibleRowAction_.flush();
+    }
+  }
+
   template <bool DoOptionalJoin>
   void runJoin() {
     while (true) {
       BlockStatus blockStatus = fillBuffer();
       if (leftSide_.currentBlocks_.empty() ||
           rightSide_.currentBlocks_.empty()) {
+        addRemainingUndefPairs();
+        if constexpr (potentiallyHasUndef) {
+          if (!rightSide_.undefBlocks_.empty()) {
+            return;
+          }
+        }
```

Comment:
you have this multiple times. It is easier if you write a function `rightHasUndef` or `hasUndef(rightSide)` to make the code shorter (the compiler should also optimize without the `potentiallyHasUndef`, and otherwise the runtime-`if` also is cheap.